aptos-move/framework/aptos-stdlib/sources/cryptography/bulletproofs.move

```diff
@@ -0,0 +1,210 @@
+module aptos_std::bulletproofs {
+    use aptos_std::pedersen;
+    use aptos_std::ristretto255::{Self, RistrettoPoint, Scalar, point_to_bytes, point_equals, scalar_to_bytes};
+    #[test_only]
+    use aptos_std::pedersen::new_commitment_from_point;
+
+    //
+    // Constants
+    //
+
+    /// The maximum range supported by the Bulletproofs library is [0, 2^{64}).
+    const MAX_RANGE_BITS : u64 = 64;
+
+    //
+    // Error codes
+    //
+
+    /// Error deserializing one of the arguments.
+    const E_DESERIALIZE_RANGE_PROOF: u64 = 1;
+
+    /// The input value given to the prover is too large.
+    const E_VALUE_OUTSIDE_RANGE: u64 = 2;
+
+    /// The range proof system only supports proving ranges of type [0, 2^{bits}) where bits \in {8, 16, 32, 64}.
+    const E_RANGE_NOT_SUPPORTED: u64 = 3;
+
+    //
+    // Structs
+    //
+
+    /// Represents a zero-knowledge range proof that a value committed inside a Pedersen commitment lies in [0, 2^{MAX_RANGE_BITS}).
+    struct RangeProof has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    //
+    // Public functions
+    //
+
+    /// Returns the maximum # of bits the range proof system can verify proofs for.
+    public fun get_max_range_bits(): u64 {
+        MAX_RANGE_BITS
+    }
+
+    /// Deserializes a range proof from a sequence of bytes.
+    public fun range_proof_from_bytes(bytes: vector<u8>): RangeProof {
+        // TODO: Can we do any pre-validation here?
+        RangeProof {
+            bytes
+        }
+    }
+
+    /// Returns the byte-representation of a range proof.
+    public fun range_proof_to_bytes(proof: &RangeProof): vector<u8> {
+        proof.bytes
+    }
+
+    #[test_only]
+    /// Computes a range proof for the Pedersen commitment to 'val' with 'randomness', unde the default Bulletproofs
+    /// commitment key; see `pedersen::new_commitment_for_bulletproof`. Returns the said commitment too.
+    ///  Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun prove_range(val: &Scalar, r: &Scalar, num_bits: u64, dst: vector<u8>): (RangeProof, pedersen::Commitment) {
+        let (bytes, compressed_comm) = prove_range_internal(scalar_to_bytes(val), scalar_to_bytes(r), num_bits, dst);
+        let point = ristretto255::new_compressed_point_from_bytes(compressed_comm);
+        let point = &std::option::extract(&mut point);
+
+        (
+            RangeProof { bytes },
+            pedersen::new_commitment_from_compressed(point)
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` committed in `com` (under the default Bulletproofs
+    /// commitment; see `pedersen::new_commitment_for_bulletproof`) satisfies $v \in [0, 2^{num_bits})$. Only works
+    /// for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof(com: &pedersen::Commitment, proof: &RangeProof, num_bits: u64, dst: vector<u8>): bool {
+        assert!(num_bits <= MAX_RANGE_BITS, std::error::invalid_argument(E_RANGE_NOT_SUPPORTED));
```

Comment:
I can still submit arbitrary numbers here, no? i.e `25` is valid input

Reply:
Yes, they will be caught in the native, actually, so that check on line 80 will be removed.

Reply:
interesting, what does the error message look like in that case? I ask because the `assert!` is as 'nice' as we get in Move- is it also easy to understand from native?

Reply:
The native will return the same abort code wrapped in `invalid_argument`! See the `bulletproofs.rs` file (screenshot below):

[](https://i.imgur.com/HW3MSvN.png)